### PR TITLE
Added support for class-string and class-string<Type> types

### PR DESCRIPTION
**Pull request by othercorey**
This adds support for parsing `class-string` and `class-string<Type>`.

### CODE REVIEW

**Comment by mvriel** [comment]
What is `class-string`? I saw a reference to it in Psalm but couldn't quickly figure out what it does. My first gut instinct says it is an alias for self; which makes me wonder why to introduce a new type.

I try to err on the safe side introducing new types as each 'special' type can cause a conflict in people's codebases.

Can you provide a bit of background for this change?

**Comment by othercorey** [comment]
This is annotation supported by many static analyzers like phpstan/psalm/etc. It identifies string types that can only represent class names.

`TypeResolver::class`

**Comment by mvriel** [comment]
I just did some research and found Psalm and PHPStan to provide quite a few more pseudo-types. This does mean we may need to re-evaluate how to deal with an increasing amount of potentially conflicting pseudo-types.

I need to check whether pseudo-types currently take precedence to the imports/uses in the Context information; because if we are to add more of these it is good to declare that an import takes precedence over a pseudo-type.

I don't have the availability to do that at this moment, I will get back to this as soon as I can

File: src/TypeResolver.php [hunk 1]
```diff
@@ -396,7 +432,7 @@ private function resolveCollection(ArrayIterator $tokens, Type $classType, Conte
         $isArray    = ((string) $classType === 'array');
         $isIterable = ((string) $classType === 'iterable');
 
-        // allow only "array", "iterable" or class name before "<"
+        // allow only "array", "iterable", "class-string" or class name before "<"
```
Comment (othercorey) [inline src/TypeResolver.php]: This is accidentally left from hacking into resolveCollection().

**Comment by orklah** [comment]
> I need to check whether pseudo-types currently take precedence to the imports/uses in the Context information; because if we are to add more of these it is good to declare that an import takes precedence over a pseudo-type.

I believe they solved this issue by adding a "-" in most pseudo types making them incompatible with actual imports.

**Comment by mvriel** [comment]
Yes, indeed. Though I am not sure whether the type resolver automatically behaves in the same way. The TypeResolver was written to be more flexible than PHP itself to better support pseudo-types :)

**Comment by mvriel** [comment]
Just checked everything and this is a-ok! Thanks for your patience

**Pull request merged**